slack_import: Support converting Slack's integration bot messages.

PieterCK opened the pull request:
### Overview
This PR adds support for handling Slack integration bot messages. Previously, this type of message wasn't supported because Slack's integration bots aren't Slack users, so we couldn't create a dummy Zulip user to send those messages during conversion.

Prep PRs:
- #32162
- #32164
- #32163

Fixes: #31162

<details>
<summary>


## Follow up:
- #31930 

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

| Slack | Zulip |
|--------|--------|
| ![image](https://github.com/user-attachments/assets/cf5ae327-fb18-47ea-a292-4e89643333fc) | ![image](https://github.com/user-attachments/assets/985e839c-e23c-4ba8-bec0-eafb311bb199) |
| ![image](https://github.com/user-attachments/assets/7553a2b6-af10-4d1b-9263-039d0ea53509) | ![image](https://github.com/user-attachments/assets/fbbdbedc-b0d7-4920-9ca9-2cc9bbfba521) |
| ![image](https://github.com/user-attachments/assets/98b7c19d-0ceb-47ea-beb5-c22e8878435f) | ![image](https://github.com/user-attachments/assets/880456d5-8d62-437f-9596-90fe810532ca) | 


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

PieterCK commented:
@kennethnrk Please review this!

@zulipbot add "buddy review"

alya commented:
@kennethnrk are you planning to review this one?

Willian-Zhang commented:
Sorry for jumping in but I tried the patch here myself but having trouble imports bot messages. I will explain in a review.

Willian-Zhang commented:
after patching the above code with this
``` python
            try:
                user = get_slack_api_data(
                    "https://slack.com/api/users.info", "user", token=token, user=user_id
                )
            except KeyError:
                # This means that the user is a bot and not a user
                print(f"Falling back to bot.info for user_id: {user_id} on mirror_dummy_user_ids, you can safely ignore this warning if it doesn't flood you.")
                bot_info = get_slack_api_data(
                    "https://slack.com/api/bots.info", "bot", token=token, bot=user_id
                )
                #TODO: previous_team_id
                bot_user = convert_bot_info_to_slack_user(bot_info, None)
                bot_user["team_domain"] = get_team_domain(team_id_to_domain, "", token)
                user_list.append(bot_user)
                added_bot_users.append(user_id)
                
                continue
            user.update(
                {
                    "team_domain": get_team_domain(team_id_to_domain, user["team_id"], token),
                    "is_mirror_dummy": True,
                }
            )
            user_list.append(user)
```

I can confirm the import is working

Willian-Zhang commented:
however I'm still having an issue that, the bot avatar are not properly imported.

PieterCK commented:
@Willian-Zhang thanks for testing the PR. I've polished the logic around how we create dummy users for integration bots, we shouldn't have any edge cases where we fail to convert them anymore. 

But I reckon adding support to convert their avatar would probably be a bit more complex because our current process is built around the `avatar_hash` (in Slack it is `hash_url`) data which the `bots.infos` API doesn't provide for integration bots, so we'd have to add a bit more logic to support converting integration bots profile which only have `image_32`data. 

I think we can focus on getting the base support for converting Slack's integration bot messages up and running first, and then build up additional features (like avatar for bots) once we've got a solid base. Or if it's a necessary requirement for you, I can add it can be added here but I reckon it'd probably introduce a bit of churn to the PR. @sbansal1999 thoughts?

user info API: https://api.slack.com/methods/users.info
bot info API: https://api.slack.com/methods/bots.info

PieterCK commented:
@zulipbot add "mentor review" I think we can do mentor review in parallel for this one. @sbansal1999

Willian-Zhang commented:
> I think we can focus on getting the base support for converting Slack's integration bot messages up and running first, and then build up additional features (like avatar for bots) once we've got a solid base. Or if it's a necessary requirement for you, I can include it here but it'd probably introduce a bit of churn to the PR. @sbansal1999 thoughts?

no its not a necessary requirement. I thought that was implemented since I haven't fully read the PR code, just patched what I had to make it not crash.

Willian-Zhang commented:
Thank you for the effort, I also have here some more crash reports and hot fixes for my other conversions. (Though they are not on the lashest commits:

PieterCK commented:
Would love to see pieces of the logs if it's something you can share, running it on a real Slack export would certainly reveal many more rough spots and edge cases that we can polish.

Willian-Zhang commented:
I have included all my hot fixes in the previous review, please check

Willian-Zhang commented:
there is also some import differences I notice between the converted conversation imported to original rendered in slack, based on the outdated hot-fixed PR:

Willian-Zhang commented:
slack uploaded images when posted in the chat directly with URL format `https://exmaple-workspace.slack.com/files/xxx/xxxxx/image.png`

Willian-Zhang commented:
sorry one more:
```
2024-08-28 14:20:06.399 ERR  [] Error processing item: [':fire:', '0/82bd772857b4529c4f8625f402b07b3d78fe62f4.png', '0/82bd772857b4529c4f8625f402b07b3d78fe62f4.original']
Traceback (most recent call last):
  File "/home/zulip/deployments/2024-08-28-14-09-08/zerver/data_import/import_util.py", line 621, in wrapping_function
    f(item)
  File "/home/zulip/deployments/2024-08-28-14-09-08/zerver/data_import/import_util.py", line 554, in get_avatar
    response = requests.get(avatar_url + size_url_suffix, stream=True)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/srv/zulip-venv-cache/95df5f42afd5755975ac90e4c8fb7527bc70bb4f/zulip-py3-venv/lib/python3.12/site-packages/requests/api.py", line 73, in get
    return request("get", url, params=params, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/srv/zulip-venv-cache/95df5f42afd5755975ac90e4c8fb7527bc70bb4f/zulip-py3-venv/lib/python3.12/site-packages/requests/api.py", line 59, in request
    return session.request(method=method, url=url, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/srv/zulip-venv-cache/95df5f42afd5755975ac90e4c8fb7527bc70bb4f/zulip-py3-venv/lib/python3.12/site-packages/requests/sessions.py", line 589, in request
    resp = self.send(prep, **send_kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/srv/zulip-venv-cache/95df5f42afd5755975ac90e4c8fb7527bc70bb4f/zulip-py3-venv/lib/python3.12/site-packages/requests/sessions.py", line 697, in send
    adapter = self.get_adapter(url=request.url)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/srv/zulip-venv-cache/95df5f42afd5755975ac90e4c8fb7527bc70bb4f/zulip-py3-venv/lib/python3.12/site-packages/requests/sessions.py", line 792, in get_adapter
    raise InvalidSchema(f"No connection adapters were found for {url!r}")
requests.exceptions.InvalidSchema: No connection adapters were found for ':fire:-512'
Stack (most recent call last):
  File "/home/zulip/deployments/2024-08-28-14-09-08/./manage.py", line 150, in <module>
    execute_from_command_line(sys.argv)
  File "/home/zulip/deployments/2024-08-28-14-09-08/./manage.py", line 115, in execute_from_command_line
    utility.execute()
  File "/srv/zulip-venv-cache/95df5f42afd5755975ac90e4c8fb7527bc70bb4f/zulip-py3-venv/lib/python3.12/site-packages/django/core/management/__init__.py", line 436, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/srv/zulip-venv-cache/95df5f42afd5755975ac90e4c8fb7527bc70bb4f/zulip-py3-venv/lib/python3.12/site-packages/django/core/management/base.py", line 413, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/home/zulip/deployments/2024-08-28-14-09-08/zerver/lib/management.py", line 97, in execute
    super().execute(*args, **options)
  File "/srv/zulip-venv-cache/95df5f42afd5755975ac90e4c8fb7527bc70bb4f/zulip-py3-venv/lib/python3.12/site-packages/django/core/management/base.py", line 459, in execute
    output = self.handle(*args, **options)
  File "/home/zulip/deployments/2024-08-28-14-09-08/zerver/management/commands/convert_slack_data.py", line 70, in handle
    do_convert_data(
  File "/home/zulip/deployments/2024-08-28-14-09-08/zerver/data_import/slack.py", line 1501, in do_convert_data
    avatar_records = process_avatars(
  File "/home/zulip/deployments/2024-08-28-14-09-08/zerver/data_import/import_util.py", line 608, in process_avatars
    run_parallel_wrapper(
  File "/home/zulip/deployments/2024-08-28-14-09-08/zerver/data_import/import_util.py", line 632, in run_parallel_wrapper
    for count, future in enumerate(
  File "/usr/lib/python3.12/concurrent/futures/_base.py", line 220, in as_completed
    fs = set(fs)
  File "/home/zulip/deployments/2024-08-28-14-09-08/zerver/data_import/import_util.py", line 633, in <genexpr>
    as_completed(executor.submit(wrapping_function, f, item) for item in full_items), 1
  File "/usr/lib/python3.12/concurrent/futures/process.py", line 831, in submit
    self._start_executor_manager_thread()
  File "/usr/lib/python3.12/concurrent/futures/process.py", line 770, in _start_executor_manager_thread
    self._launch_processes()
  File "/usr/lib/python3.12/concurrent/futures/process.py", line 797, in _launch_processes
    self._spawn_process()
  File "/usr/lib/python3.12/concurrent/futures/process.py", line 807, in _spawn_process
    p.start()
  File "/usr/lib/python3.12/multiprocessing/process.py", line 121, in start
    self._popen = self._Popen(self)
  File "/usr/lib/python3.12/multiprocessing/context.py", line 282, in _Popen
    return Popen(process_obj)
  File "/usr/lib/python3.12/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(process_obj)
  File "/usr/lib/python3.12/multiprocessing/popen_fork.py", line 71, in _launch
    code = process_obj._bootstrap(parent_sentinel=child_r)
  File "/usr/lib/python3.12/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/usr/lib/python3.12/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python3.12/concurrent/futures/process.py", line 263, in _process_worker
    r = call_item.fn(*call_item.args, **call_item.kwargs)
  File "/home/zulip/deployments/2024-08-28-14-09-08/zerver/data_import/import_util.py", line 623, in wrapping_function
    logging.exception("Error processing item: %s", item, stack_info=True)
```

Willian-Zhang commented:
another hotfix taped my problem:

```python
import emoji_data_python

def get_emoji_avatar(id):
    # example https://github.com/iamcal/emoji-data/blob/master/img-apple-160/2665-fe0f.png?raw=true
    name = id[1:-1]
    char = emoji_data_python.emoji_short_names[name]
    code = char.variations[0] if len(char.variations) > 0 else char.unified
    code = code.lower()
    url = f"https://github.com/iamcal/emoji-data/blob/master/img-apple-160/{code}.png?raw=true"
    return requests.get(url, stream=True)
    
    
def get_avatar(avatar_dir: str, size_url_suffix: str, avatar_upload_item: list[str]) -> None:
    avatar_url = avatar_upload_item[0]

    image_path = os.path.join(avatar_dir, avatar_upload_item[1])
    original_image_path = os.path.join(avatar_dir, avatar_upload_item[2])
    if avatar_url[0] == avatar_url[-1] == ':':
        print(f"using emoji avatar {avatar_url}")
        response = get_emoji_avatar(avatar_url)
    else:
        response = requests.get(avatar_url + size_url_suffix, stream=True)
    with open(image_path, "wb") as image_file:
        shutil.copyfileobj(response.raw, image_file)
    shutil.copy(image_path, original_image_path)
```

note: please do not use this in production as the URL included here is not meant to be used for large scale requests

Willian-Zhang commented:
another issue while importing, I'm not sure if I should post it here but it might be an Mal-converstion issue:

```python
2024-08-28 18:05:55.325 INFO [] Successfully rendered Markdown for message batch
Traceback (most recent call last):
  File "/home/zulip/deployments/2024-08-03-00-42-06/./manage.py", line 150, in <module>
    execute_from_command_line(sys.argv)
  File "/home/zulip/deployments/2024-08-03-00-42-06/./manage.py", line 115, in execute_from_command_line
    utility.execute()
  File "/srv/zulip-venv-cache/58c4cec1a70fbcdbd716f3aa98fe57c891b4ee84/zulip-py3-venv/lib/python3.10/site-packages/django/core/management/__init__.py", line 436, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/srv/zulip-venv-cache/58c4cec1a70fbcdbd716f3aa98fe57c891b4ee84/zulip-py3-venv/lib/python3.10/site-packages/django/core/management/base.py", line 413, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/home/zulip/deployments/2024-08-03-00-42-06/zerver/lib/management.py", line 97, in execute
    super().execute(*args, **options)
  File "/srv/zulip-venv-cache/58c4cec1a70fbcdbd716f3aa98fe57c891b4ee84/zulip-py3-venv/lib/python3.10/site-packages/django/core/management/base.py", line 459, in execute
    output = self.handle(*args, **options)
  File "/home/zulip/deployments/2024-08-03-00-42-06/zerver/management/commands/import.py", line 107, in handle
    do_import_realm(path, subdomain, num_processes)
  File "/home/zulip/deployments/2024-08-03-00-42-06/zerver/lib/import_realm.py", line 1496, in do_import_realm
    import_message_data(realm=realm, sender_map=sender_map, import_dir=import_dir)
  File "/home/zulip/deployments/2024-08-03-00-42-06/zerver/lib/import_realm.py", line 1734, in import_message_data
    bulk_import_model(data, Message)
  File "/home/zulip/deployments/2024-08-03-00-42-06/zerver/lib/import_realm.py", line 742, in bulk_import_model
    model.objects.bulk_create(model(**item) for item in data[table])
  File "/srv/zulip-venv-cache/58c4cec1a70fbcdbd716f3aa98fe57c891b4ee84/zulip-py3-venv/lib/python3.10/site-packages/django/db/models/manager.py", line 87, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/srv/zulip-venv-cache/58c4cec1a70fbcdbd716f3aa98fe57c891b4ee84/zulip-py3-venv/lib/python3.10/site-packages/django/db/models/query.py", line 818, in bulk_create
    returned_columns = self._batched_insert(
  File "/srv/zulip-venv-cache/58c4cec1a70fbcdbd716f3aa98fe57c891b4ee84/zulip-py3-venv/lib/python3.10/site-packages/django/db/models/query.py", line 1875, in _batched_insert
    self._insert(
  File "/srv/zulip-venv-cache/58c4cec1a70fbcdbd716f3aa98fe57c891b4ee84/zulip-py3-venv/lib/python3.10/site-packages/django/db/models/query.py", line 1847, in _insert
    return query.get_compiler(using=using).execute_sql(returning_fields)
  File "/srv/zulip-venv-cache/58c4cec1a70fbcdbd716f3aa98fe57c891b4ee84/zulip-py3-venv/lib/python3.10/site-packages/django/db/models/sql/compiler.py", line 1823, in execute_sql
    cursor.execute(sql, params)
  File "/srv/zulip-venv-cache/58c4cec1a70fbcdbd716f3aa98fe57c891b4ee84/zulip-py3-venv/lib/python3.10/site-packages/django/db/backends/utils.py", line 79, in execute
    return self._execute_with_wrappers(
  File "/srv/zulip-venv-cache/58c4cec1a70fbcdbd716f3aa98fe57c891b4ee84/zulip-py3-venv/lib/python3.10/site-packages/django/db/backends/utils.py", line 92, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/srv/zulip-venv-cache/58c4cec1a70fbcdbd716f3aa98fe57c891b4ee84/zulip-py3-venv/lib/python3.10/site-packages/django/db/backends/utils.py", line 105, in _execute
    return self.cursor.execute(sql, params)
  File "/home/zulip/deployments/2024-08-03-00-42-06/zerver/lib/db.py", line 39, in execute
    wrapper_execute(self, super().execute, query, vars)
  File "/home/zulip/deployments/2024-08-03-00-42-06/zerver/lib/db.py", line 22, in wrapper_execute
    action(sql, params)
ValueError: A string literal cannot contain NUL (0x00) characters.
```

Willian-Zhang commented:
identified content that caused this.
```
"content": "`\u0000\u0001\u0001\u0001\u0001\u0000\u0000\u0000\u0001\u0000\u0000\u0001/Users/blankwonder/Workspaces/Surge/Surge-Core/Connector\u0000/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/clang/7.0.0/include\u0000/Users/blankwonder/Workspaces/Surge/Surge-iOS/Vendor/openssl/include/openssl\u0000/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS9.0.sdk/usr/include/_types\u0000\u0000SSEncrypt.c\u0000\u0001\u0000\u0000stddef.h\u0000\u0002\u0000\u0000evp.h\u0000\u0003\u0000\u0000asn1.h\u0000\u0003\u0000\u0000ossl_typ.h\u0000\u0003\u0000\u0000_uint64_t.h\u0000\u0004\u0000\u0000SSEncrypt.h\u0000\u0001\u0000\u0000\u0000\u0000`",
```

alya commented:
@PieterCK are you continuing work on this one?

@mateuszmandera should be able to review or help out if needed.

PieterCK commented:
Ah, sorry I lost track of this issue. I've cleaned up the PR a bit and addressed some comments. 

## Emoji avatar crash:
https://github.com/zulip/zulip/pull/31311#issuecomment-2315526663 
hmm, our `process_avatar` function tried to process an avatar profile with ":fire:" as the avatar_url here. I'm not sure how it got an emoji string as the avatar URL. 

I ended up adding a new commit to properly convert PNG avatar URLs (Slacks integration bots only have these) and default users with unexpected avatar_urls to use basic gravatar URLs instead. commit https://github.com/zulip/zulip/pull/31311/commits/b6990f39492d82f3c15afc9f5ded7d8a3ae8cf4f. 

Integration bot avatars should now be able to be imported and users with unknown avatars (e.g., emoji) should not crash the importer, they'll have default gravatar avatars. 

## NUL Byte crash:
https://github.com/zulip/zulip/pull/31311#issuecomment-2316010058
Yes, I was also able to replicate this by inserting that message into my raw Slack export and importing it. It looks like the codes in this PR weren't the ones causing it as I tried importing with an older version. In that case, our other third-party importer might also be susceptible.
<details>
<summary>Slack message containing NUL bytes</summary>

```
{
        "user": "U077VQAAAQL",
        "type": "message",
        "ts": "1718207226.487079",
        "client_msg_id": "bfca851b-5246-4dec-9549-32ad66a6d463",
        "text": "\u0000\u0001\u0001\u0001\u0001\u0000\u0000\u0000\u0001\u0000\u0000\u0001/Users/blankwonder/Workspaces/Surge/Surge-Core/Connector\u0000/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/clang/7.0.0/include\u0000/Users/blankwonder/Workspaces/Surge/Surge-iOS/Vendor/openssl/include/openssl\u0000/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS9.0.sdk/usr/include/_types\u0000\u0000SSEncrypt.c\u0000\u0001\u0000\u0000stddef.h\u0000\u0002\u0000\u0000evp.h\u0000\u0003\u0000\u0000asn1.h\u0000\u0003\u0000\u0000ossl_typ.h\u0000\u0003\u0000\u0000_uint64_t.h\u0000\u0004\u0000\u0000SSEncrypt.h\u0000\u0001\u0000\u0000\u0000\u0000",
        "team": "T06NRA6HM3P",
        "user_team": "T06NRA6HM3P",
        "source_team": "T06NRA6HM3P",
        "user_profile": {
            "avatar_hash": "2642ad691131",
            "image_72": "https:\/\/avatars.slack-edge.com\/2024-06-12\/7265349337474_2642ad69113105b829c3_72.jpg",
            "first_name": "Satyam",
            "real_name": "Satyam Bansal",
            "display_name": "Satyam Bansal",
            "team": "T06NRA6HM3P",
            "name": "sbansal1999",
            "is_restricted": false,
            "is_ultra_restricted": false
        },
        "blocks": [
            {
                "type": "rich_text",
                "block_id": "wSisM",
                "elements": [
                    {
                        "type": "rich_text_section",
                        "elements": [
                            {
                                "type": "text",
                                "text": "testing slack side of things"
                            }
                        ]
                    }
                ]
            }
        ]
    }
```
</details>

<details>
<summary>error log</summary>

```
(zulip-py3-venv) vagrant@d680cd97b1d3:/srv/zulip$ python manage.py import "dssasds" /tmp/converted-slack-data-sg95h0u9/
Processing dump: /tmp/converted-slack-data-sg95h0u9 ...
2024-10-09 13:34:08.194 INFO [] Importing realm dump /tmp/converted-slack-data-sg95h0u9
2024-10-09 13:34:08.195 INFO [] Importing realm data from /tmp/converted-slack-data-sg95h0u9/realm.json
2024-10-09 13:34:08.294 INFO [] Successfully imported <class 'zerver.models.streams.Stream'> from zerver_stream.
2024-10-09 13:34:08.311 WARN [] Dropped restricted authentication method: AzureAD
2024-10-09 13:34:08.312 WARN [] Dropped restricted authentication method: SAML
2024-10-09 13:34:08.314 INFO [] Successfully imported <class 'zerver.models.realms.RealmAuthenticationMethod'> from zerver_realmauthenticationmethod.
2024-10-09 13:34:08.316 INFO [] Successfully imported <class 'zerver.models.streams.DefaultStream'> from zerver_defaultstream.
2024-10-09 13:34:08.318 INFO [] Successfully imported <class 'zerver.models.realm_emoji.RealmEmoji'> from zerver_realmemoji.
2024-10-09 13:34:08.320 INFO [] Successfully imported <class 'zerver.models.realms.RealmDomain'> from zerver_realmdomain.
2024-10-09 13:34:08.322 INFO [] Successfully imported <class 'zerver.models.linkifiers.RealmFilter'> from zerver_realmfilter.
2024-10-09 13:34:08.323 INFO [] Successfully imported <class 'zerver.models.realm_playgrounds.RealmPlayground'> from zerver_realmplayground.
2024-10-09 13:34:08.343 INFO [] Successfully imported <class 'zerver.models.recipients.Recipient'> from zerver_recipient.
2024-10-09 13:34:08.353 INFO [] Successfully imported <class 'zerver.models.streams.Subscription'> from zerver_subscription.
2024-10-09 13:34:08.353 INFO [] about to call create_subscription_events
2024-10-09 13:34:08.370 INFO [] done with create_subscription_events
2024-10-09 13:34:08.372 INFO [] Successfully imported <class 'zerver.models.recipients.DirectMessageGroup'> from zerver_huddle.
2024-10-09 13:34:08.391 INFO [] Successfully imported <class 'zerver.models.presence.UserPresence'> from zerver_userpresence.
2024-10-09 13:34:08.392 INFO [] Successfully imported <class 'zerver.models.user_activity.UserActivity'> from zerver_useractivity.
2024-10-09 13:34:08.393 INFO [] Successfully imported <class 'zerver.models.user_activity.UserActivityInterval'> from zerver_useractivityinterval.
2024-10-09 13:34:08.395 INFO [] Successfully imported <class 'zerver.models.custom_profile_fields.CustomProfileField'> from zerver_customprofilefield.
2024-10-09 13:34:08.396 INFO [] Successfully imported <class 'zerver.models.custom_profile_fields.CustomProfileFieldValue'> from zerver_customprofilefieldvalue.
2024-10-09 13:34:08.396 INFO [] Importing avatars
2024-10-09 13:34:08.570 INFO [] Importing uploaded files
2024-10-09 13:34:08.570 INFO [] Importing emojis
2024-10-09 13:34:08.770 INFO [] Importing realm icons and logos
2024-10-09 13:34:08.771 INFO [] Importing message dump /tmp/converted-slack-data-sg95h0u9/messages-000001.json
2024-10-09 13:34:08.867 INFO [] Successfully rendered Markdown for message batch
Traceback (most recent call last):
  File "/srv/zulip/manage.py", line 150, in <module>
    execute_from_command_line(sys.argv)
  File "/srv/zulip/manage.py", line 115, in execute_from_command_line
    utility.execute()
  File "/srv/zulip-venv-cache/b477df627afb33e3ce1cd594561fcbb9d3946a14/zulip-py3-venv/lib/python3.10/site-packages/django/core/management/__init__.py", line 436, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/srv/zulip-venv-cache/b477df627afb33e3ce1cd594561fcbb9d3946a14/zulip-py3-venv/lib/python3.10/site-packages/django/core/management/base.py", line 413, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/srv/zulip/zerver/lib/management.py", line 97, in execute
    super().execute(*args, **options)
  File "/srv/zulip-venv-cache/b477df627afb33e3ce1cd594561fcbb9d3946a14/zulip-py3-venv/lib/python3.10/site-packages/django/core/management/base.py", line 459, in execute
    output = self.handle(*args, **options)
  File "/srv/zulip/zerver/management/commands/import.py", line 107, in handle
    do_import_realm(path, subdomain, num_processes)
  File "/srv/zulip/zerver/lib/import_realm.py", line 1616, in do_import_realm
    import_message_data(realm=realm, sender_map=sender_map, import_dir=import_dir)
  File "/srv/zulip/zerver/lib/import_realm.py", line 1854, in import_message_data
    bulk_import_model(data, Message)
  File "/srv/zulip/zerver/lib/import_realm.py", line 745, in bulk_import_model
    model.objects.bulk_create(model(**item) for item in data[table])
  File "/srv/zulip-venv-cache/b477df627afb33e3ce1cd594561fcbb9d3946a14/zulip-py3-venv/lib/python3.10/site-packages/django/db/models/manager.py", line 87, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/srv/zulip-venv-cache/b477df627afb33e3ce1cd594561fcbb9d3946a14/zulip-py3-venv/lib/python3.10/site-packages/django/db/models/query.py", line 818, in bulk_create
    returned_columns = self._batched_insert(
  File "/srv/zulip-venv-cache/b477df627afb33e3ce1cd594561fcbb9d3946a14/zulip-py3-venv/lib/python3.10/site-packages/django/db/models/query.py", line 1875, in _batched_insert
    self._insert(
  File "/srv/zulip-venv-cache/b477df627afb33e3ce1cd594561fcbb9d3946a14/zulip-py3-venv/lib/python3.10/site-packages/django/db/models/query.py", line 1847, in _insert
    return query.get_compiler(using=using).execute_sql(returning_fields)
  File "/srv/zulip-venv-cache/b477df627afb33e3ce1cd594561fcbb9d3946a14/zulip-py3-venv/lib/python3.10/site-packages/django/db/models/sql/compiler.py", line 1823, in execute_sql
    cursor.execute(sql, params)
  File "/srv/zulip-venv-cache/b477df627afb33e3ce1cd594561fcbb9d3946a14/zulip-py3-venv/lib/python3.10/site-packages/django/db/backends/utils.py", line 122, in execute
    return super().execute(sql, params)
  File "/srv/zulip-venv-cache/b477df627afb33e3ce1cd594561fcbb9d3946a14/zulip-py3-venv/lib/python3.10/site-packages/django/db/backends/utils.py", line 79, in execute
    return self._execute_with_wrappers(
  File "/srv/zulip-venv-cache/b477df627afb33e3ce1cd594561fcbb9d3946a14/zulip-py3-venv/lib/python3.10/site-packages/django/db/backends/utils.py", line 92, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/srv/zulip-venv-cache/b477df627afb33e3ce1cd594561fcbb9d3946a14/zulip-py3-venv/lib/python3.10/site-packages/django/db/backends/utils.py", line 105, in _execute
    return self.cursor.execute(sql, params)
  File "/srv/zulip/zerver/lib/db.py", line 39, in execute
    wrapper_execute(self, super().execute, query, vars)
  File "/srv/zulip/zerver/lib/db.py", line 22, in wrapper_execute
    action(sql, params)
ValueError: A string literal cannot contain NUL (0x00) characters.
(zulip-py3-venv) vagrant@d680cd97b1d3:/srv/zulip$ 

```
</details>
I've filed  #31930 to track this. Will check other importer and report it to that issue, for now. Let me know if we want to include this in this PR.

## Message with file format:
https://github.com/zulip/zulip/pull/31311#issuecomment-2312696590
Yes, Slack file URLs aren't publicly available by default, so if formatted like shown (hyperlink) it'll not have any preview. It looks like the file attachment is exported properly and I wasn't able to reproduce the repeated formatting.

I'm not sure if there are any practical things we can do to remove the repeated file attachment message for this case. Is the message from one of your integrations by any chance? 

<details>

<summary>example of message with file attachments</summary>

slack:
![image](https://github.com/user-attachments/assets/c6b83e9c-f63c-4f71-a6e8-96c1aace2053)

imported:
![image](https://github.com/user-attachments/assets/300ddc85-d602-4a5c-8dc0-4b5fb5559316)

```
[
    {
        "text": "",
        "files": [
            {
                "id": "F07R0NC7YJZ",
                "created": 1728479093,
                "timestamp": 1728479093,
                "name": "7057208497908_a4351f6deb91094eac4c_512.png",
                "title": "7057208497908_a4351f6deb91094eac4c_512.png",
                "mimetype": "image\/png",
                "filetype": "png",
                "pretty_type": "PNG",
                "user": "U06NU4E26M9",
                "user_team": "T06NRA6HM3P",
                "editable": false,
                "size": 50202,
                "mode": "hosted",
                "is_external": false,
                "external_type": "",
                "is_public": true,
                "public_url_shared": false,
                "display_as_bot": false,
                "username": "",
                "url_private": "https:\/\/files.slack.com\/files-pri\/T06NRA6HM3P-F07R0NC7YJZ\/7057208497908_a4351f6deb91094eac4c_512.png?t=xoxe-6773346599125-7866680305409-7853905921843-0afb929289a606016aa941efe2427a2d",
                "url_private_download": "https:\/\/files.slack.com\/files-pri\/T06NRA6HM3P-F07R0NC7YJZ\/download\/7057208497908_a4351f6deb91094eac4c_512.png?t=xoxe-6773346599125-7866680305409-7853905921843-0afb929289a606016aa941efe2427a2d",
                "media_display_type": "unknown",
                "thumb_64": "https:\/\/files.slack.com\/files-tmb\/T06NRA6HM3P-F07R0NC7YJZ-5f5dd4dcf9\/7057208497908_a4351f6deb91094eac4c_512_64.png?t=xoxe-6773346599125-7866680305409-7853905921843-0afb929289a606016aa941efe2427a2d",
                "thumb_80": "https:\/\/files.slack.com\/files-tmb\/T06NRA6HM3P-F07R0NC7YJZ-5f5dd4dcf9\/7057208497908_a4351f6deb91094eac4c_512_80.png?t=xoxe-6773346599125-7866680305409-7853905921843-0afb929289a606016aa941efe2427a2d",
                "thumb_360": "https:\/\/files.slack.com\/files-tmb\/T06NRA6HM3P-F07R0NC7YJZ-5f5dd4dcf9\/7057208497908_a4351f6deb91094eac4c_512_360.png?t=xoxe-6773346599125-7866680305409-7853905921843-0afb929289a606016aa941efe2427a2d",
                "thumb_360_w": 360,
                "thumb_360_h": 360,
                "thumb_480": "https:\/\/files.slack.com\/files-tmb\/T06NRA6HM3P-F07R0NC7YJZ-5f5dd4dcf9\/7057208497908_a4351f6deb91094eac4c_512_480.png?t=xoxe-6773346599125-7866680305409-7853905921843-0afb929289a606016aa941efe2427a2d",
                "thumb_480_w": 480,
                "thumb_480_h": 480,
                "thumb_160": "https:\/\/files.slack.com\/files-tmb\/T06NRA6HM3P-F07R0NC7YJZ-5f5dd4dcf9\/7057208497908_a4351f6deb91094eac4c_512_160.png?t=xoxe-6773346599125-7866680305409-7853905921843-0afb929289a606016aa941efe2427a2d",
                "original_w": 512,
                "original_h": 512,
                "thumb_tiny": "AwAwADDToopkkixrlvwHrQCVxzMFGWOBQrBhkHIqi8jStk\/gKfG5Q8fiKTdjX2ehcopEYOMilpmQVTuYXMu4ZYH9KuU1pFQgMwGelNNrYqMnF3I4YBGMnlv5U6SIPyOG\/nUlFJ6hzO9yGJGDZPGKmpFZWzg5xS0lHl0E3fUKoXMTq5cksp7+lX6KuMuViKMdw6DHDD3pzTu4xwB7VO1vGTnGPpQsEa9s\/Wr5o72E0yOBGLBugHerNFFZt3BKx\/\/Z",
                "permalink": "https:\/\/ds-py62195.slack.com\/files\/U06NU4E26M9\/F07R0NC7YJZ\/7057208497908_a4351f6deb91094eac4c_512.png",
                "permalink_public": "https:\/\/slack-files.com\/T06NRA6HM3P-F07R0NC7YJZ-13e18f29a1",
                "is_starred": false,
                "has_rich_preview": false,
                "file_access": "visible"
            }
        ],
        "upload": false,
        "user": "U06NU4E26M9",
        "display_as_bot": false,
        "type": "message",
        "ts": "1728479096.997459",
        "client_msg_id": "21be3bc8-d519-4b4b-83a1-f5335a59020d"
    },
    {
        "text": "<@U074VRHQ11T>",
        "files": [
            {
                "id": "F07RGKLAZ8R",
                "created": 1728482387,
                "timestamp": 1728482387,
                "name": "pabbly.svg",
                "title": "pabbly.svg",
                "mimetype": "image\/svg+xml",
                "filetype": "svg",
                "pretty_type": "SVG",
                "user": "U06NU4E26M9",
                "user_team": "T06NRA6HM3P",
                "editable": false,
                "size": 2492,
                "mode": "hosted",
                "is_external": false,
                "external_type": "",
                "is_public": true,
                "public_url_shared": false,
                "display_as_bot": false,
                "username": "",
                "url_private": "https:\/\/files.slack.com\/files-pri\/T06NRA6HM3P-F07RGKLAZ8R\/pabbly.svg?t=xoxe-6773346599125-7866680305409-7853905921843-0afb929289a606016aa941efe2427a2d",
                "url_private_download": "https:\/\/files.slack.com\/files-pri\/T06NRA6HM3P-F07RGKLAZ8R\/download\/pabbly.svg?t=xoxe-6773346599125-7866680305409-7853905921843-0afb929289a606016aa941efe2427a2d",
                "media_display_type": "unknown",
                "permalink": "https:\/\/ds-py62195.slack.com\/files\/U06NU4E26M9\/F07RGKLAZ8R\/pabbly.svg",
                "permalink_public": "https:\/\/slack-files.com\/T06NRA6HM3P-F07RGKLAZ8R-ab33d2ac26",
                "is_starred": false,
                "has_rich_preview": false,
                "file_access": "visible"
            }
        ],
        "upload": false,
        "user": "U06NU4E26M9",
        "display_as_bot": false,
        "blocks": [
            {
                "type": "rich_text",
                "block_id": "+hPn2",
                "elements": [
                    {
                        "type": "rich_text_section",
                        "elements": [
                            {
                                "type": "user",
                                "user_id": "U074VRHQ11T"
                            }
                        ]
                    }
                ]
            }
        ],
        "type": "message",
        "ts": "1728482390.074489",
        "client_msg_id": "381909be-c505-4820-ac4a-eacfad11125c"
    }
]
```
</details>

PieterCK commented:
As for the text formatting quirks, for the most part, I think those are known quirks from our Slack text conversion...

@mateuszmandera- Could you please review this PR? Thank you

PieterCK commented:
@zulipbot remove "buddy review"

Willian-Zhang commented:
> Message with file format:
> https://github.com/zulip/zulip/pull/31311#issuecomment-2312696590
> Yes, Slack file URLs aren't publicly available by default, so if formatted like shown (hyperlink) it'll not have any preview. It looks like the file attachment is exported properly and I wasn't able to reproduce the repeated formatting.
>
> I'm not sure if there are any practical things we can do to remove the repeated file attachment message for this case. Is the message from one of your integrations by any chance?

they are not, it turn out to be, form what I can see two cases:
1. messages that people upload image and post directly became similar two image links, one with token one without.
2. quoted (slack archive) messages that contained preview of image that only has the nontokened ones

I've created my own script keying all known attachments and images by name and timestamp (to int precision) and match them with all links, successfully clear all such issues

PieterCK commented:
>they are not, it turn out to be, form what I can see two cases:
>
>   1. messages that people upload image and post directly became similar two image links, one with token one without.
>   2. quoted (slack archive) messages that contained preview of image that only has the nontokened ones
> ...

Ahh I see, that's interesting. I'll pull our discussion here to this CZO [discussion](https://chat.zulip.org/#narrow/stream/3-backend/topic/Slack.20importer.20discussion.20from.20.2331311) so that other folks can chip in. Thank you for the report!

Willian-Zhang commented:
> > "content": "`\u0000\u0001\u0001\u0001\u0001\u0000\u0000\u0000\u0001\u0000\u0000\u0001/Users/blankwonder/Workspaces/Surge/Surge-Core/Connector\u0000/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../lib/clang/7.0.0/include\u0000/Users/blankwonder/Workspaces/Surge/Surge-iOS/Vendor/openssl/include/openssl\u0000/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS9.0.sdk/usr/include/_types\u0000\u0000SSEncrypt.c\u0000\u0001\u0000\u0000stddef.h\u0000\u0002\u0000\u0000evp.h\u0000\u0003\u0000\u0000asn1.h\u0000\u0003\u0000\u0000ossl_typ.h\u0000\u0003\u0000\u0000_uint64_t.h\u0000\u0004\u0000\u0000SSEncrypt.h\u0000\u0001\u0000\u0000\u0000\u0000`",
> 
> 
> 
> @Willian-Zhang - So this message is from the export after you've converted the Slack export right? I wonder whether the original Slack exported message looks identical to this or not.., I couldn't reproduce an export data containing this kind of forbidden string, if I copy and paste the string above directly to my Slack workspace and try to export it, Slack export will character escape these string so that it can be safely handled when imported. I'm suspect that the text reformatting function I added in this PR might have unexpected side effect to specific format of string, so that'd be great if we could confirm whether that's the case or not. 
> 
> 
> 
> <details>
> 
> 
> 
> Original form Slack:
> 
> ![image](https://github.com/user-attachments/assets/6823aaf4-c4b8-401a-a87a-7649e5dccfef)
> 
> 
> 
> Once exported, Slack will convert them to safe form. This can be imported:
> 
> >"text": "````\\u0000\\u0001\\u0001\\u0001\\u0001\\u0000\\u0000\\u0000\\u0001\\u0000\\u0000\\u0001\/Users\/blankwonder\/Workspaces\/Surge\/Surge-Core\/Connector\\u0000\/Applications\/Xcode.app\/Contents\/Developer\/Toolchains\/XcodeDefault.xctoolchain\/usr\/bin\/..\/lib\/clang\/7.0.0\/include\\u0000\/Users\/blankwonder\/Workspaces\/Surge\/Surge-iOS\/Vendor\/openssl\/include\/openssl\\u0000\/Applications\/Xcode.app\/Contents\/Developer\/Platforms\/iPhoneOS.platform\/Developer\/SDKs\/iPhoneOS9.0.sdk\/usr\/include\/_types\\u0000\\u0000SSEncrypt.c\\u0000\\u0001\\u0000\\u0000stddef.h\\u0000\\u0002\\u0000\\u0000evp.h\\u0000\\u0003\\u0000\\u0000asn1.h\\u0000\\u0003\\u0000\\u0000ossl_typ.h\\u0000\\u0003\\u0000\\u0000_uint64_t.h\\u0000\\u0004\\u0000\\u0000SSEncrypt.h\\u0000\\u0001\\u0000\\u0000\\u0000\\u0000````\n",
> 
> 
> 
> Imported, in Zulip:
> 
> ![image](https://github.com/user-attachments/assets/cf217ccf-1d8e-44ed-8555-51a362906546)
> 
> 
> 
> </details>

The original message in the slack should not be encoded like what's in the json literal, please try decode it before pasting into slack.
They should show up indistinguishable as **space** as of last time I checked them on slack.

PieterCK commented on zerver/tests/test_slack_importer.py:
```diff
@@ -540,7 +610,9 @@ def test_users_to_zerver_userprofile(self, mock_get_data_file: mock.Mock) -> Non
                 customprofilefield_value,
-            ) = users_to_zerver_userprofile(slack_data_dir, user_data, 1, timestamp, "test_domain")
+            ) = users_to_zerver_userprofile(
+                slack_data_dir, user_data, 1, timestamp, "testdomain.com"
```
this is changed because in #31795 we added a check to make sure converted user emails are valid

alya commented:
@laurynmm would it make sense for you to take a look here?

PieterCK commented on zerver/data_import/slack_message_conversion.py:
```diff
@@ -215,10 +217,12 @@ def replace_links(text: str) -> str:
     )
-    if block_type == "actions":
-        # Unhandled
+    unhandled_types = ["actions", "rich_text"]
```
"rich_text" block type is basically the same as the "text" field, so this is filtered out to avoid duplication

PieterCK commented:
@laurynmm Thanks for checking this out! I tried breaking out the commits in this PR into separate PRs. This is my first time "stacking" PRs like this, so feel free to let me know if it’s not what you had in mind. Most of them still need work, but the first one (marked with "STAGE=0") should be ready for review.
- #32162 (Ready for review)
- #32163 (DRAFT)
- #32164 (DRAFT)
- #32165 (DRAFT)

PieterCK commented on zerver/data_import/slack.py:
```diff
@@ -931,8 +932,8 @@ def channel_message_to_zerver_message(
             continue
-
         try:
+            message["text"] += process_slack_block_and_attachment(message)
```
`process_slack_block_and_attachment` doesn't do any text reformatting, so we pass its output through `convert_to_zulip_markdown` too.

PieterCK replied:
And moved to the `try` block so that the importer doesn't crash over very rare edge cases

PieterCK commented:
This PR is clear of any blocker, previously it was pending on #32164. I've rebased this and manually tested by importing Slack export with integration bot formatting, posted the result in the description.

PieterCK commented:
@zulipbot add "area: export/import"

alya commented:
As noted on CZO, adding before/after screenshots of the *same* message (and possibly several, if that would showcase different formatting) will help the review process. Using a table is a good way to present them.

PieterCK commented on zerver/data_import/slack.py:
```diff
@@ -932,6 +933,22 @@ def channel_message_to_zerver_message(
+            # Leave it as is if formatted_block is an empty string, it's likely
+            # one of the unhandled_types.
+            if formatted_block != "":
+                message["text"] = formatted_block
```
we do the same thing in Slack incoming webhook -- override message["text"].

PieterCK commented on zerver/tests/test_slack_importer.py:
```diff
@@ -1446,6 +1446,184 @@ def test_channel_message_to_zerver_message_with_threads(
+        self.assertEqual(zerver_message[1]["content"], expected_message_block)
+        self.assertEqual(zerver_message[1]["sender"], slack_user_id_to_zulip_user_id["B06NWMNUQ3W"])
+
+        expected_message_block_2 = """## [Buy Ingredients!](https://app.clickup.com/t/86cuy8e4y?comment=90160026391938)
```
Extensive tests for `render_blocks` and `render_attachments` is already done in `slack_incoming/test.py`, this just tests that it works well as part of the import tool

timabbott replied:
Let's put that into a comment/docstring at the top of this test, something like:

> Most of the core logic for converting Slack blocks is in `slack_incoming/test.py`, so the purpose for this test is to verify the import tool portion of this system.

timabbott commented on zerver/data_import/slack.py:
```diff
@@ -932,6 +933,24 @@ def channel_message_to_zerver_message(
+                # expect.
+                formatted_block = process_slack_block_and_attachment(message)
+            except Exception:
+                print("Slack message has unexpected component format.")
```
Let's not eat the exception -- we want this to be debuggable.

PieterCK replied:
I also removed this condition:
```python
      if "blocks" in message or "attachments" in message:
          try:
           ...
```
since all messages wil have the "blocks" components

timabbott commented on zerver/data_import/slack_message_conversion.py:
```diff
@@ -239,10 +241,16 @@ def convert_to_zulip_markdown(
+
+    unhandled_types = [
+        "actions",
+        "rich_text",  # All messages sent by users will have this format
```
You mean "All block messages", right? Also put the comment on its own line above this one; inline comments are annoying.

timabbott replied:
Probably useful would also be for both of these unhandled types fo have comments on the line before them explaining what roughly would be required to handle them, for posterity.

PieterCK replied:
Updated it to:
```python
    unhandled_types = [
        # The "actions" block is used to format literal in-message clickable
        # buttons and similar elements, which Zulip currently doesn't support.
        # https://docs.slack.dev/reference/block-kit/blocks/actions-block
        "actions",
        # All user-sent messages contain at least a "block" component with a
        # "rich_text" block. This block contains the same string as the "text"
        # field. We're skipping this because the Slack import tool already
        # handles the "text" field and the Slack incoming integration
        # overrides it.
        # https://docs.slack.dev/reference/block-kit/blocks/rich-text-block/
        "rich_text",
    ]
```

timabbott commented on zerver/data_import/slack_message_conversion.py:
```diff
@@ -389,3 +397,15 @@ def replace_links(text: str) -> str:
+
+    if slack_message.get("attachments"):
+        pieces += map(render_attachment, slack_message["attachments"])
+    return "\n\n".join(piece.strip() for piece in pieces if piece.strip() != "")
```
Do we need to join with double-newlines?

PieterCK replied:
hmm, yeah I'd think so.

I ran the slack incoming tests with a single line of newline and here's a sample:
```
======================================================================
FAIL: test_complicated (zerver.webhooks.slack_incoming.tests.SlackIncomingHookTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/srv/zulip/zerver/webhooks/slack_incoming/tests.py", line 171, in test_complicated
    self.check_webhook(
  File "/srv/zulip/zerver/lib/test_classes.py", line 2460, in check_webhook
    self.assert_channel_message(
  File "/srv/zulip/zerver/lib/test_classes.py", line 2476, in assert_channel_message
    self.assertEqual(message.content, content)
  File "/srv/zulip/zerver/lib/test_classes.py", line 295, in assertEqual
    raise AssertionError(
AssertionError: Actual and expected outputs do not match; showing diff.
  ## Hello from TaskBot
+ 
  Hey there 👋 I'm TaskBot. I'm here to help you create and manage tasks in Slack.
  There are two ways to quickly create tasks:
+ 
  **1️⃣ Use the `/task` command**. Type `/task` followed by a short description of your tasks and I'll ask for a due date (if applicable). Try it out by using the `/task` command in this channel.
+ 
  **2️⃣ Use the *Create a Task* action.** If you want to create a task from a message, select `Create a Task` in a message's context menu. Try it out by selecting the *Create a Task* action for this message (shown below).
+ 
  [image1](https://api.slack.com/img/blocks/bkb_template_images/onboardingComplex.jpg)
+ 
  ➕ To start tracking your team's tasks, **add me to a channel** and I'll introduce myself. I'm usually added to a team or project channel. Type `/invite @TaskBot` from the channel or pick a channel on the right.
+ 
  ----
+ 
  [cute cat](https://pbs.twimg.com/profile_images/625633822235693056/lNGUneLX_400x400.jpg)
  
  👀 View all tasks with `/task list`
  
  ❓Get help at any time with:
  - `/task help`, or
  - type **help** in a DM with me
```

output:
```quote
## Hello from TaskBot
Hey there 👋 I'm TaskBot. I'm here to help you create and manage tasks in Slack.
There are two ways to quickly create tasks:
**1️⃣ Use the `/task` command**. Type `/task` followed by a short description of your tasks and I'll ask for a due date (if applicable). Try it out by using the `/task` command in this channel.
**2️⃣ Use the *Create a Task* action.** If you want to create a task from a message, select `Create a Task` in a message's context menu. Try it out by selecting the *Create a Task* action for this message (shown below).
[image1](https://api.slack.com/img/blocks/bkb_template_images/onboardingComplex.jpg)
➕ To start tracking your team's tasks, **add me to a channel** and I'll introduce myself. I'm usually added to a team or project channel. Type `/invite @TaskBot` from the channel or pick a channel on the right.
----
[cute cat](https://pbs.twimg.com/profile_images/625633822235693056/lNGUneLX_400x400.jpg)
👀 View all tasks with `/task list`
❓Get help at any time with:
- `/task help`, or
- type **help** in a DM with me
```

I  think one of the main reason for the double newlines is to keep converted images from stacking with the other texts and other images:
![image](https://github.com/user-attachments/assets/fbe1b394-ac60-484f-9a41-61f57a36931c)

PieterCK commented:
Update: Previously CI failed because I left in some changes in `slack_incoming` when testing. Rolled back the change there.

timabbott commented:
Can we do a newline just after images?

PieterCK commented:
| Before | Only "image" have double `\n` |
|--------|--------|
| ![image](https://github.com/user-attachments/assets/c81dfc23-e947-456b-8049-ea9aea5c5858) | ![image](https://github.com/user-attachments/assets/cf1448db-6450-4db5-9506-dc6e5ed6070d) |
| ![image](https://github.com/user-attachments/assets/99317756-7614-4a2b-b656-47996b3eeb21) | ![image](https://github.com/user-attachments/assets/5fc0da77-8647-4752-923f-4e12b98d22bf) |

I feel like it's good for payloads with multiple "sections" block containing somewhat long text (the first 3 paragraphs in the first example), but it can look awkward for short "sections" blocks, say, only 1-2 line each like the second examples. Sorry for not posting a complete example before, should I open a #design discussion for this?

alya commented:
"Before" seems easier to read to me.

PieterCK commented:
One can play around with Slack's blocks/attachments component here: [Block Kit Builder](https://app.slack.com/block-kit-builder/T06NRA6HM3P#%7B%22blocks%22:%5B%7B%22type%22:%22section%22,%22text%22:%7B%22type%22:%22mrkdwn%22,%22text%22:%22Danny%20Torrence%20left%20the%20following%20review%20for%20your%20property:%22%7D%7D,%7B%22type%22:%22section%22,%22block_id%22:%22section567%22,%22text%22:%7B%22type%22:%22mrkdwn%22,%22text%22:%22%3Chttps://example.com%7COverlook%20Hotel%3E%20%5Cn%20:star:%20%5Cn%20Doors%20had%20too%20many%20axe%20holes,%20guest%20in%20room%20237%20was%20far%20too%20rowdy,%20whole%20place%20felt%20stuck%20in%20the%201920s.%22%7D,%22accessory%22:%7B%22type%22:%22image%22,%22image_url%22:%22https://is5-ssl.mzstatic.com/image/thumb/Purple3/v4/d3/72/5c/d3725c8f-c642-5d69-1904-aa36e4297885/source/256x256bb.jpg%22,%22alt_text%22:%22Haunted%20hotel%20image%22%7D%7D,%7B%22type%22:%22section%22,%22block_id%22:%22section789%22,%22fields%22:%5B%7B%22type%22:%22mrkdwn%22,%22text%22:%22*Average%20Rating*%5Cn1.0%22%7D%5D%7D%5D%7D).
I've updated https://github.com/zulip/zulip/pull/31311#issuecomment-2885653938 to include screenshots of how the same payloads would be rendered by Slack

timabbott commented:
I agree "before" looks better, and more similar to the Slack original. Let's go with that approach. Does that require any changes?

PieterCK commented:
> I agree "before" looks better, and more similar to the Slack original. Let's go with that approach. Does that require any changes?

Nope, that doesn't require any changes. "Before" is the current state. And I just ran a manual test again just to confirm.

timabbott commented:
Merged, thanks for all the work on this @PieterCK!
